lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,423 @@
+
+The operator needs to handle the failure case using one of the available
+[failure-handling features][fail] in VRL.
+
+#### Type Checking
+
+In addition to expressions being fallible, the type checker also considers
+a program fallible if the type expected by an expression cannot be guaranteed at
+compile-time.
+
+For example:
+
+```coffee
+.message = "log message: " + .log
+```
+
+In this case, the `log` field type cannot be determined at compile-time, and
+thus concatenating the field value with a string might fail (e.g. if it's an
+array, or any other type that cannot be combined with a string).
+
+This too needs to be handled at compile-time by the operator.
+
+#### Progressive Type Checking
+
+A quirk in the type checker is that arguments passed to functions are _not_ type
+checked individually. Instead, the function call itself can be marked fallible
+if any of its arguments do not adhere to the expected type.
+
+For example:
+
+```coffee
+upcase(.message)
+```
+
+Upcasing a string is an infallible operation, however, because we can't
+guarantee that the `message` field will actually be a string, the function call
+is still invalid, as the function is marked as fallible.
+
+This decision [was made][#6507] for ergonomics purposes.
+
+[fail]: https://vrl.dev/errors/#runtime-errors
+[#6507]: https://github.com/vectordotdev/vector/issues/6507
+
+### Errors
+
+- All errors are caught at compile-time by the compiler (see "fallibility"
+  chapter).
+
+- The only exception to this rule is if the operator explicitly allows
+  a function to fail the program at runtime (e.g. `safe_call()` vs
+  `unsafe_call!()`).
+
+- A function should be marked as "fallible" if its internal implementation can
+  fail.
+
+- A function _should not_ be marked as fallible if it receives the wrong
+  argument type. This is handled by the compiler.
+
+- Errors should contain explicit messages detailing what went wrong, and how the
+  operator can solve the problem.
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If usability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Functions should be preceded with their function category for organization and
+  discovery.
+
+  - Use `parse_*` for functions that decode a string to another type of data
+    (e.g. `parse_json` and `parse_grok`).
+  
+  - Use `decode_*` for string to string decoding functions (e.g.
+    `decode_base64`).
+  
+  - Use `encode_*` for string encoding functions (e.g. `encode_base64`).
```

Comment:
I'm still partial to the `unmarshal` / `marshal` terminology for converting from/to serialized structured data (like `unmarshal_json`); leaving `encode` / `decode` for converting bytes to each other, but I'm ok if we want to leave that discussion for later. What you have here does codify the existing patterns.

Reply:
Created https://github.com/vectordotdev/vector/issues/9405 to track this.